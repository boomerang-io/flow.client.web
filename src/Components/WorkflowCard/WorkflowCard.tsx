import React, { useState } from "react";
import axios from "axios";
import { useAppContext } from "Hooks";
import { useFeature } from "flagged";
import { useMutation, queryCache } from "react-query";
import { Link, useHistory } from "react-router-dom";
import {
  Button,
  ConfirmModal,
  ComposedModal,
  InlineLoading,
  OverflowMenu,
  OverflowMenuItem,
  ToastNotification,
  TooltipIcon,
  notify,
} from "@boomerang-io/carbon-addons-boomerang-react";
import WorkflowWarningButton from "Components/WorkflowWarningButton";
import UpdateWorkflow from "./UpdateWorkflow";
import WorkflowInputModalContent from "./WorkflowInputModalContent";
import WorkflowRunModalContent from "./WorkflowRunModalContent";
import fileDownload from "js-file-download";
import { formatErrorMessage } from "@boomerang-io/utils";
import { appLink, FeatureFlag } from "Config/appConfig";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { BASE_URL } from "Config/servicesConfig";
import { Run20, Bee20 } from "@carbon/icons-react";
import workflowIcons from "Assets/workflowIcons";
import { ComposedModalChildProps, FlowTeamQuotas, ModalTriggerProps, WorkflowSummary } from "Types";
import { WorkflowScope } from "Constants";
import styles from "./workflowCard.module.scss";

interface WorkflowCardProps {
  scope: string;
  teamId: string | null;
  quotas: FlowTeamQuotas | null;
  workflow: WorkflowSummary;
}

type FunctionAnyReturn = () => any;

const WorkflowCard: React.FC<WorkflowCardProps> = ({ scope, teamId, quotas, workflow }) => {
  const { teams } = useAppContext();
  const cancelRequestRef = React.useRef<FunctionAnyReturn | null>();
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
  const [isUpdateWorkflowModalOpen, setIsUpdateWorkflowModalOpen] = useState(false);
  const workflowQuotasEnabled = useFeature(FeatureFlag.WorkflowQuotasEnabled);
  const activityEnabled = useFeature(FeatureFlag.ActivityEnabled);

  const history = useHistory();
  const [errorMessage, seterrorMessage] = useState(null);

  const [deleteWorkflowMutator, { isLoading: isDeleting }] = useMutation(resolver.deleteWorkflow, {});

  const [executeWorkflowMutator, { error: executeError, isLoading: isExecuting }] = useMutation(
    (args: { id: string; properties: {} }) => {
      const { promise, cancel } = resolver.postExecuteWorkflow(args);
      cancelRequestRef.current = cancel;
      return promise;
    }
  );

  const [duplicateWorkflowMutator, { isLoading: duplicateWorkflowIsLoading }] = useMutation(
    resolver.postDuplicateWorkflow
  );

  const isDuplicating = duplicateWorkflowIsLoading;

  /**
   * Format properties to be edited in form by Formik. It doesn't work with property notation :(
   * See: https://jaredpalmer.com/formik/docs/guides/arrays#nested-objects
   * This is safe to do because we don't accept "-" characters in property keys
   * @returns {Array}
   */
  const formatPropertiesForEdit = () => {
    const { properties = [] } = workflow;
    return properties.filter((property) => !property.readOnly);
  };

  const handleDeleteWorkflow = async () => {
    const workflowId = workflow.id;
    try {
      await deleteWorkflowMutator({ id: workflowId });
      notify(<ToastNotification kind="success" title="Delete Workflow" subtitle="Workflow successfully deleted" />);
      if (scope === WorkflowScope.Team) {
        /**
         * teams query takes a while. optomistic update here
         */
        const specificTeam = teams.find((team) => team.id === teamId);
        const specificTeamIndex = teams.findIndex((team) => team.id === teamId);
        const newTeamWorkflows = specificTeam?.workflows.filter((workflow) => workflow.id !== workflowId);
        // @ts-ignore
        teams[specificTeamIndex].workflows = newTeamWorkflows;
        queryCache.setQueryData(serviceUrl.getTeams(), teams);
        queryCache.invalidateQueries(serviceUrl.getTeams());
      } else if (scope === WorkflowScope.System) {
        queryCache.invalidateQueries(serviceUrl.getSystemWorkflows());
      } else {
        queryCache.invalidateQueries(serviceUrl.getUserWorkflows());
      }
    } catch {
      notify(<ToastNotification kind="error" title="Something's Wrong" subtitle="Request to delete workflow failed" />);
    }
  };

  const handleDuplicateWorkflow = async (workflow: WorkflowSummary) => {
    try {
      await duplicateWorkflowMutator({ workflowId: workflow.id });
      notify(
        <ToastNotification kind="success" title="Duplicate Workflow" subtitle="Successfully duplicated workflow" />
      );
      if (scope === WorkflowScope.System) {
        queryCache.invalidateQueries(serviceUrl.getSystemWorkflows());
      } else if (scope === WorkflowScope.Team) {
        queryCache.invalidateQueries(serviceUrl.getTeams());
      } else {
        queryCache.invalidateQueries(serviceUrl.getUserWorkflows());
      }
      return;
    } catch (e) {
      notify(
        <ToastNotification kind="error" title="Something's Wrong" subtitle="Request to duplicate workflow failed" />
      );
      return;
    }
  };

  const handleExportWorkflow = (workflow: WorkflowSummary) => {
    notify(<ToastNotification kind="info" title="Export Workflow" subtitle="Export starting soon" />);
    axios
      .get(`${BASE_URL}/workflow/export/${workflow.id}`)
      .then(({ data }) => {
        fileDownload(JSON.stringify(data, null, 4), `${workflow.name}.json`);
      })
      .catch((error) => {
        notify(<ToastNotification kind="error" title="Something's Wrong" subtitle="Export workflow failed" />);
      });
  };

  const handleExecuteWorkflow = async (closeModal: () => void, redirect: boolean = false, properties: {} = {}) => {
    const { id: workflowId } = workflow;
    try {
      const { data: execution } = await executeWorkflowMutator({ id: workflowId, properties });
      notify(
        <ToastNotification kind="success" title="Run Workflow" subtitle="Successfully started workflow execution" />
      );
      if (redirect) {
        history.push({
          pathname: appLink.execution({ executionId: execution.id, workflowId }),
          state: { fromUrl: appLink.workflows(), fromText: "Workflows" },
        });
      } else {
        closeModal();
      }
    } catch (err) {
      seterrorMessage(
        formatErrorMessage({
          error: err,
          defaultMessage: "Run Workflow Failed",
        })
      );
      //no-op
    }
  };

  let menuOptions = [
    {
      itemText: "Edit",
      onClick: () => history.push(appLink.editorDesigner({ workflowId: workflow.id })),
    },
    {
      itemText: "View Activity",
      onClick: () => history.push(appLink.workflowActivity({ workflowId: workflow.id })),
    },
    {
      itemText: "Update",
      onClick: () => setIsUpdateWorkflowModalOpen(true),
    },
    {
      itemText: "Export",
      onClick: () => handleExportWorkflow(workflow),
    },
    {
      itemText: "Duplicate",
      onClick: () => handleDuplicateWorkflow(workflow),
    },
    {
      hasDivider: true,
      itemText: "Delete",
      isDelete: true,
      onClick: () => setIsDeleteModalOpen(true),
    },
  ];

  if (!activityEnabled) {
    menuOptions = menuOptions.filter((el) => el.itemText !== "View Activity");
  }

  const formattedProperties = formatPropertiesForEdit();

  const { name, Icon = Bee20 } = workflowIcons.find((icon) => icon.name === workflow.icon) ?? {};

  let hasReachedMonthlyRunLimit = false;

  if (quotas) {
    hasReachedMonthlyRunLimit = quotas?.maxWorkflowExecutionMonthly <= quotas?.currentWorkflowExecutionMonthly;
  }

  const canRunManually = workflow?.triggers?.manual?.enable ?? false;

  const isDisabled = workflowQuotasEnabled && (hasReachedMonthlyRunLimit || !canRunManually);

  let loadingText = "";

  if (isExecuting) {
    loadingText = "Executing...";
  } else if (isDuplicating) {
    loadingText = "Duplicating...";
  } else if (isDeleting) {
    loadingText = "Deleting...";
  }

  return (
    <div className={styles.container}>
      <Link to={!isDeleting ? appLink.editorDesigner({ workflowId: workflow.id }) : ""}>
        <section className={styles.details}>
          <div className={styles.iconContainer}>
            <Icon className={styles.icon} alt={`${name}`} />
          </div>
          <div className={styles.descriptionContainer}>
            <h1 title={workflow.name} className={styles.name} data-testid="workflow-card-title">
              {workflow.name}
            </h1>
            <p title={workflow.shortDescription} className={styles.description}>
              {workflow.shortDescription}
            </p>
          </div>
        </section>
      </Link>
      {/*<div className={styles.quotaDescriptionContainer}>
        <h3
          className={styles.teamQuotaText}
        >{`Run quota - ${quotas.currentWorkflowExecutionMonthly} of ${quotas.maxWorkflowExecutionMonthly} this month`}</h3>
        {hasReachedMonthlyRunLimit && (
          <TooltipHover
            direction="top"
            tooltipText="This Workflow has reached the maximum number of executions allowed this month. Contact your Team administrator/owner to increase the quota, or wait until the quota resets next month."
          >
            <WarningAlt16 className={styles.warningIcon} />
          </TooltipHover>
        )}
        </div>*/}
      <section className={styles.launch}>
        {Array.isArray(formattedProperties) && formattedProperties.length !== 0 ? (
          <ComposedModal
            modalHeaderProps={{
              title: "Workflow Parameters",
              subtitle: "Provide parameter values for your workflow",
            }}
            modalTrigger={({ openModal }: ModalTriggerProps) => (
              <Button
                disabled={isDeleting || isDisabled}
                iconDescription="Run Workflow"
                renderIcon={Run20}
                size="field"
                onClick={openModal}
              >
                Run it
              </Button>
            )}
            onCloseModal={() => {
              if (cancelRequestRef.current) cancelRequestRef.current();
            }}
          >
            {({ closeModal }: ComposedModalChildProps) => (
              <WorkflowInputModalContent
                closeModal={closeModal}
                executeError={executeError}
                executeWorkflow={handleExecuteWorkflow}
                isExecuting={isExecuting}
                inputs={formattedProperties}
              />
            )}
          </ComposedModal>
        ) : (
          <ComposedModal
            composedModalProps={{ containerClassName: `${styles.executeWorkflow}` }}
            modalHeaderProps={{
              title: "Execute Workflow",
              subtitle: '"Run and View" will navigate you to the workflow exeuction view.',
            }}
            modalTrigger={({ openModal }: ModalTriggerProps) => (
              <Button
                disabled={isDeleting || isDisabled}
                iconDescription="Run Workflow"
                renderIcon={Run20}
                size="field"
                onClick={openModal}
              >
                Run it
              </Button>
            )}
          >
            {({ closeModal }: ComposedModalChildProps) => (
              <WorkflowRunModalContent
                closeModal={closeModal}
                executeError={executeError}
                executeWorkflow={handleExecuteWorkflow}
                isExecuting={isExecuting}
                errorMessage={errorMessage}
              />
            )}
          </ComposedModal>
        )}
      </section>
      {workflow.templateUpgradesAvailable && (
        <div className={styles.templatesWarningIcon}>
          <TooltipIcon direction="top" tooltipText={"New version of a task available! To update, edit your workflow."}>
            <WorkflowWarningButton />
          </TooltipIcon>
        </div>
      )}
      {isDuplicating || isDeleting || isExecuting ? (
        <InlineLoading
          description={loadingText}
          style={{ position: "absolute", right: "0.5rem", top: "0", width: "fit-content" }}
        />
      ) : (
        <OverflowMenu
          flipped
          ariaLabel="Overflow card menu"
          iconDescription="Overflow menu icon"
          style={{ position: "absolute", right: "0" }}
        >
          {menuOptions.map(({ onClick, itemText, ...rest }, index) => (
            <OverflowMenuItem onClick={onClick} itemText={itemText} key={`${itemText}-${index}`} {...rest} />
          ))}
        </OverflowMenu>
      )}
      {isUpdateWorkflowModalOpen && (
        <UpdateWorkflow
          onCloseModal={() => setIsUpdateWorkflowModalOpen(false)}
          teamId={teamId}
          workflowId={workflow.id}
        />
      )}
      {isDeleteModalOpen && (
        <ConfirmModal
          affirmativeAction={handleDeleteWorkflow}
          affirmativeButtonProps={{ kind: "danger" }}
          affirmativeText="Delete"
          isOpen={isDeleteModalOpen}
          negativeAction={() => {
            setIsDeleteModalOpen(false);
          }}
          negativeText="Cancel"
          onCloseModal={() => {
            setIsDeleteModalOpen(false);
          }}
          title="Delete Workflow"
        >
          Are you sure you want to delete this workflow? There's no going back from this decision.
        </ConfirmModal>
      )}
    </div>
  );
};

export default WorkflowCard;
