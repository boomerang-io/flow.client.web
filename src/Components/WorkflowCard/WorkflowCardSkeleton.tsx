import { SkeletonPlaceholder } from "@carbon/react";

import styles from "./workflowCard.module.scss";

const WorkflowCardSkeleton = () => {
  return <SkeletonPlaceholder className={styles.skeletonContainer}></SkeletonPlaceholder>;
};

export default WorkflowCardSkeleton;
