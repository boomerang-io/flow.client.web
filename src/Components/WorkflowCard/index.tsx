export { default } from "./WorkflowCard";
export { default as WorkflowCardSkeleton } from "./WorkflowCardSkeleton";
