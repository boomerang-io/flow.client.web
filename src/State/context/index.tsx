import React from "react";
import { FlowTeam, FlowUser, TaskModel, WorkflowRevision, WorkflowSummary } from "Types";

export function createContext<ContextType>() {
  const context = React.createContext<ContextType | undefined>(undefined);
  function useContext() {
    const contextValue = React.useContext(context);
    if (!contextValue) throw new Error("useContext must be inside a Provider with a value");
    return contextValue;
  }
  return [useContext, context.Provider] as const;
}

export const [useAppContext, AppContextProvider] = createContext<AppContext>();

type AppContext = {
  isTutorialActive: boolean;
  communityUrl: string;
  setIsTutorialActive: (isActive: boolean) => void;
  user: FlowUser;
  teams: FlowTeam[];
};

interface taskProvider {
  category: string;
  id: string;
  icon: any;
  name: string;
}

interface stepInterface {
  activityId: string;
  duration: number;
  flowTaskStatus: string;
  id: string;
  order: number;
  outputs: any;
  startTime: string;
  taskId: string;
  taskName: string;
  runWorkflowActivityStatus: string;
}

interface workflowExecutionInterface {
  creationDate: string;
  duration: number;
  id: string;
  properties: any; //array, may need to be changed later
  status: string;
  steps: Array<stepInterface>;
  teamName: string;
  trigger: string;
  workflowId: string;
  workflowRevisionid: string;
}

interface ExecutionContext {
  tasks: Array<taskProvider>;
  workflowExecution: workflowExecutionInterface;
  workflowRevision: object;
}
export const [useExecutionContext, ExecutionContextProvider] = createContext<ExecutionContext>();

interface EditorContext {
  revisionDispatch?: Function;
  revisionState: WorkflowRevision;
  summaryData: WorkflowSummary;
  taskTemplatesData: Array<TaskModel>;
}

export const [useEditorContext, EditorContextProvider] = createContext<EditorContext>();
