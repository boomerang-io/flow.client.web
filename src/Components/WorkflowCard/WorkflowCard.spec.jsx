import React from "react";
import { startApiServer } from "ApiServer";
import { teams, profile } from "ApiServer/fixtures";
import { AppContextProvider } from "State/context";
import WorkflowCard from "./index";

const props = {
  teamId: teams[0].id,
  quotas: teams[0].workflowQuotas,
  workflow: teams[0].workflows,
};

let server;

beforeEach(() => {
  server = startApiServer({ environment: "test" });
});

afterEach(() => {
  server.shutdown();
});

describe("WorkflowCard --- Snapshot", () => {
  it("Capturing Snapshot of WorkflowCard", () => {
    const { baseElement } = rtlContextRouterRender(
      <AppContextProvider
        value={{
          isTutorialActive: false,
          setIsTutorialActive: () => {},
          user: profile,
          teams,
        }}
      >
        <WorkflowCard {...props} />
      </AppContextProvider>
    );
    expect(baseElement).toMatchSnapshot();
  });
});
