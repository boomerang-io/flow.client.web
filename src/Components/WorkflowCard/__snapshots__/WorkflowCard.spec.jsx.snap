// Vitest Snapshot v1

exports[`WorkflowCard --- Snapshot > Capturing Snapshot of WorkflowCard 1`] = `
<body
  id="app"
>
  <div>
    <div>
      <a
        href="/editor/undefined/workflow"
      >
        <section>
          <div>
            <svg
              alt="undefined"
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="20"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 10a6 6 0 00-6 6v8a6 6 0 0012 0V16A6 6 0 0016 10zm-4.25 7.87h8.5v4.25h-8.5zM16 28.25A4.27 4.27 0 0111.75 24v-.13h8.5V24A4.27 4.27 0 0116 28.25zm4.25-12.13h-8.5V16a4.25 4.25 0 018.5 0zM30.66 19.21L24 13v9.1a4 4 0 008 0A3.83 3.83 0 0030.66 19.21zM28 24.35a2.25 2.25 0 01-2.25-2.25V17l3.72 3.47h0A2.05 2.05 0 0130.2 22 2.25 2.25 0 0128 24.35zM0 22.1a4 4 0 008 0V13L1.34 19.21A3.88 3.88 0 000 22.1zm2.48-1.56h0L6.25 17v5.1a2.25 2.25 0 01-4.5 0A2.05 2.05 0 012.48 20.54zM15 5.5A3.5 3.5 0 1011.5 9 3.5 3.5 0 0015 5.5zm-5.25 0A1.75 1.75 0 1111.5 7.25 1.77 1.77 0 019.75 5.5zM20.5 2A3.5 3.5 0 1024 5.5 3.5 3.5 0 0020.5 2zm0 5.25A1.75 1.75 0 1122.25 5.5 1.77 1.77 0 0120.5 7.25z"
              />
            </svg>
          </div>
          <div>
            <h1
              data-testid="workflow-card-title"
            />
            <p />
          </div>
        </section>
      </a>
      <section>
        <button
          class="bx--btn bx--btn--md bx--btn--primary bx--btn--disabled"
          disabled=""
          tabindex="0"
          type="button"
        >
          Run it
          <svg
            aria-hidden="true"
            aria-label="Run Workflow"
            class="bx--btn__icon"
            fill="currentColor"
            focusable="false"
            height="20"
            preserveAspectRatio="xMidYMid meet"
            role="img"
            viewBox="0 0 32 32"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21,16a6,6,0,1,1-6,6,6,6,0,0,1,6-6m0-2a8,8,0,1,0,8,8,8,8,0,0,0-8-8Z"
            />
            <path
              d="M26,4H6A2,2,0,0,0,4,6V26a2,2,0,0,0,2,2h4V26H6V12H28V6A2,2,0,0,0,26,4ZM6,10V6H26v4Z"
            />
            <path
              d="M19 19L19 25 24 22 19 19z"
            />
          </svg>
        </button>
      </section>
      <button
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="Overflow card menu"
        class="bx--overflow-menu bx--overflow-menu--md"
        style="position: absolute; right: 0px;"
        type="button"
      >
        <svg
          aria-label="Overflow menu icon"
          class="bx--overflow-menu__icon"
          fill="currentColor"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          role="img"
          viewBox="0 0 32 32"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="16"
            cy="8"
            r="2"
          />
          <circle
            cx="16"
            cy="16"
            r="2"
          />
          <circle
            cx="16"
            cy="24"
            r="2"
          />
          <title>
            Overflow menu icon
          </title>
        </svg>
      </button>
    </div>
  </div>
  <div
    class="bx--bmrg-modal-portal"
  />
</body>
`;
